src/dotty/tools/dotc/printing/Formatting.scala

```diff
@@ -66,17 +69,41 @@ object Formatting {
    *  message composition methods, this is crucial.
    */
   class ErrorMessageFormatter(sc: StringContext) extends StringFormatter(sc) {
+    override protected def showArg(arg: Any)(implicit ctx: Context): String =
+      wrapNonSensical(arg, super.showArg(arg))
+  }
+
+  class SyntaxFormatter(sc: StringContext) extends StringFormatter(sc) {
     override protected def showArg(arg: Any)(implicit ctx: Context): String = {
-      def isSensical(arg: Any): Boolean = arg match {
-        case tpe: Type =>
-          tpe.exists && !tpe.isErroneous
-        case sym: Symbol if sym.isCompleted =>
-          sym.info != ErrorType && sym.info != TypeAlias(ErrorType) && sym.info.exists
-        case _ => true
+      if (ctx.settings.color.value != "never") arg match {
+        case arg: Showable =>
+          val highlighted =
+            SyntaxHighlighting(wrapNonSensical(arg, super.showArg(arg)))
+          new String(highlighted.toArray)
+        case hl: Highlight =>
+          hl.show
+        case hb: HighlightBuffer =>
+          hb.toString
+        case str: String =>
+          new String(SyntaxHighlighting(str).toArray)
+        case _ => super.showArg(arg)
       }
-      val str = super.showArg(arg)
-      if (isSensical(arg)) str else Diagnostic.nonSensicalStartTag + str + Diagnostic.nonSensicalEndTag
+      else super.showArg(arg)
+    }
+  }
+
+  private def wrapNonSensical(arg: Any, str: String)(implicit ctx: Context): String = {
```

Comment:
Semi-OT: should one establish an annotation (taking the Dotty type as a `String`)?